src/core/net/srp_client.cpp

```diff
@@ -0,0 +1,1559 @@
+    SuccessOrExit(error = mSocket.Connect(aServerSockAddr));
+
+    otLogInfoSrp("[client] Starting, server [%s]:%d", aServerSockAddr.GetAddress().ToString().AsCString(),
+                 aServerSockAddr.mPort);
+
+    mCallback        = aCallback;
+    mCallbackContext = aContext;
+    Resume();
+
+exit:
+    return error;
+}
+
+void Client::Stop(void)
+{
+    // Change the state of host info and services so that they are
+    // added/removed again once the client is started back. In the
+    // case of `kAdding`, we intentionally move to `kToRefresh`
+    // instead of `kToAdd` since the server may receive our add
+    // request and the item may be registered on the server. This
+    // ensures that if we are later asked to remove the item, we do
+    // notify server.
+
+    static const ItemState kNewStateOnStop[]{
+        /* (0) kToAdd      -> */ kToAdd,
+        /* (1) kAdding     -> */ kToRefresh,
+        /* (2) kToRefresh  -> */ kToRefresh,
+        /* (3) kRefreshing -> */ kToRefresh,
+        /* (4) kToRemove   -> */ kToRemove,
+        /* (5) kRemoving   -> */ kToRemove,
+        /* (6) kRegistered -> */ kToRefresh,
+        /* (7) kRemoved    -> */ kRemoved,
+    };
+
+    VerifyOrExit(GetState() != kStateStopped);
+
+    // State changes:
+    //   kAdding     -> kToRefresh
+    //   kRefreshing -> kToRefresh
+    //   kRemoving   -> kToRemove
+    //   kRegistered -> kToRefresh
+
+    mHostInfo.SetState(kNewStateOnStop[mHostInfo.GetState()]);
+
+    for (Service *service = mServices.GetHead(); service != nullptr; service = service->GetNext())
+    {
+        service->SetState(kNewStateOnStop[service->GetState()]);
+    }
+
+    IgnoreError(mSocket.Close());
+    mShouldRemoveKeyLease = false;
+    mTxFailureRetryCount  = 0;
+    ResetRetryWaitInterval();
+    SetState(kStateStopped);
+
+exit:
+    return;
+}
+
+void Client::Resume(void)
+{
+    SetState(kStateUpdated);
+    UpdateState();
+}
+
+void Client::Pause(void)
+{
+    // Change the state of host info and services that are are being
+    // added or removed so that they are added/removed again once the
+    // client is resumed or started back.
+
+    static const ItemState kNewStateOnPause[]{
+        /* (0) kToAdd      -> */ kToAdd,
+        /* (1) kAdding     -> */ kToRefresh,
+        /* (2) kToRefresh  -> */ kToRefresh,
+        /* (3) kRefreshing -> */ kToRefresh,
+        /* (4) kToRemove   -> */ kToRemove,
+        /* (5) kRemoving   -> */ kToRemove,
+        /* (6) kRegistered -> */ kRegistered,
+        /* (7) kRemoved    -> */ kRemoved,
+    };
+
+    // State changes:
+    //   kAdding     -> kToRefresh
+    //   kRefreshing -> kToRefresh
+    //   kRemoving   -> kToRemove
+
+    mHostInfo.SetState(kNewStateOnPause[mHostInfo.GetState()]);
+
+    for (Service *service = mServices.GetHead(); service != nullptr; service = service->GetNext())
+    {
+        service->SetState(kNewStateOnPause[service->GetState()]);
+    }
```

Comment:
```suggestion
    ChangeToNewStates(kNewStateOnPause);
```

Encapsulate this logic with a method since this logic is used in multiple places?